### PR TITLE
save as draft

### PR DESCRIPTION
**Pull request by maria-j-k**
closing #3272

### CODE REVIEW

File: app/controllers/backoffice/services/offers_controller.rb [hunk 1]
```diff
@@ -34,8 +44,15 @@ def edit
...
   def update
+    save_as_draft = params[:commit] == "Save as Draft"
     template = permitted_attributes(Offer)
-    if Offer::Update.call(@offer, transform_attributes(template, @service))
+
+    if save_as_draft
+      template[:name] = params["name"]
+      Offer::UpdateAsDraft.call(@offer, transform_attributes(template, @service))
+      redirect_to backoffice_service_path(@service), notice: "Offer updated successfully and drafted"
+      nil
```
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/controllers/backoffice/services/offers_controller.rb:54]: added by rubocop...

File: app/controllers/backoffice/services/offers_controller.rb [hunk 2]
```diff
@@ -70,6 +87,11 @@ def offer_template
     Offer.new(temp.merge(service: @service, status: :published))
   end
 
+  def offer_draft_template
+    temp = transform_attributes(permitted_attributes(Offer), @service)
```
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/controllers/backoffice/services/offers_controller.rb:91]: it was made to be consistant with the existing offer_template function

File: app/inputs/custom_input.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+class CustomInput < SimpleForm::Inputs::StringInput
```
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/inputs/custom_input.rb:2]: I've found it as a way to use in simple form based on model a custom field not related to the model

File: app/models/offer.rb [hunk 4]
```diff
@@ -44,6 +44,7 @@ def should_index?
   belongs_to :primary_oms, class_name: "OMS", optional: true
   has_many :project_items, dependent: :restrict_with_error
 
+  before_validation :set_iid
```
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/models/offer.rb:47]: We need this line for "regular" offer creation

File: app/models/offer.rb [hunk 5]
```diff
@@ -56,17 +57,16 @@ def should_index?
...
-  validate :set_iid, on: :create
   validates :service, presence: true
   validates :iid, presence: true, numericality: true
   validates :order_url, mp_url: true, if: :order_url?
 
   validate :primary_oms_exists?, if: -> { primary_oms_id.present? }
-  validate :proper_oms?, if: -> { primary_oms.present? }
-  validates :oms_params, absence: true, if: -> { current_oms.blank? }
-  validate :check_oms_params, if: -> { current_oms.present? }
+  validate :proper_oms?, if: -> { primary_oms.present? && !draft? }
```
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/models/offer.rb:65]: this validate=false in create didn't work recursivelly in 100%

**Comment by agpul** [comment]
image
I think that this box should be just above the exit button (now I can't see on my laptop). I had to change the size of the page to see the box.

**Comment by agpul** [comment]
Minor: 
1. I  entered to the offer (it had the name "Agnieszka")
2. I didn't change anything
3. I had to Save changes although I didn't change anything

**Comment by agpul** [comment]
Issue, to consider: @AlicjaSwierad 
1. I entered the offer that was Publish only to see it!!
2. I didn't change anything
3. I click "Save as draft"
4. Then I entered the offer one more time because I wanted to Publish it and I could find the Publish button (it was in the end of the form, this is NOT intuitve

File: app/models/offer.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (maria-j-k) [reply app/models/offer.rb:47]: it was used as validator
The only change is to make it a hook before validation - for consistency

**Review by goreck888 — CHANGES_REQUESTED** [review]
There are a few things you can improve to make this code more maintainable.

**Comment by AlicjaSwierad** [comment]
1. The offer drafting basic functionality is working correctly; however, there are some styling issues, particularly with the colors. 

2. I drafted the offer, which was published. When I tried to publish it again, it wasn't possible because some fields that are mandatory were empty (I hadn't changed anything)

3. There is a **major issue related to identifying the required fields within the offer**, which (I guess) is expected to be fixed in a separate PR. 

4. Is it possible to display the popup only when changes have been made? in existing offer?

**Review by goreck888 — APPROVED** [review]
Just add a changelog entry :)

**Comment by AlicjaSwierad** [comment]
Please apply styles from the file: https://www.figma.com/file/0F53l8h6P2Q7TpzLuFZ7qX/Offers%3A-Adding-and-Editing?type=design&node-id=557%3A13882&mode=design&t=dtSZerrGZ7tgwB8i-1

- [ ] Add an overlay with #000000 at 40% opacity

- [ ] Style the popup according to the design

- [ ] Fix the issue with the popup location, right now, it appears in the bottom right corner

@jarekzet

**Comment by jarekzet** [comment]
added mods in exit modal display (positioning and styling):

image

**Review by AlicjaSwierad — APPROVED** [review]

**Pull request merged**